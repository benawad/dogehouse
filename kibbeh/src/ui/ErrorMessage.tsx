import * as React from "react";
import { SmSolidPlus } from "../icons";

export interface ErrorMessageProps {
  message?: string;
  button?: string;
  autoClose?: boolean;
}

export const ErrorMessage: React.FC<ErrorMessageProps> = ({
  message = "An error has occurred",
  button = "Refresh",
  autoClose = false,
  ...props
}) => {
  const [closed, setClosed] = React.useState<boolean>(true);

  React.useEffect(() => {
    setClosed(false);
  }, []);

  React.useEffect(() => {
    if (closed || !autoClose) {
      return;
    }

    const timer = setTimeout(() => {
      setClosed(true);
    }, 7000);

    return () => {
      clearTimeout(timer);
    };
  }, [autoClose, closed]);

  return (
    <div
      className={`rounded-t-8 w-full fixed items-center justify-center text-button transition-transform duration-300`}
      style={{
        height: 50,
        backgroundColor: "#5575E7",
        bottom: 0,
        left: 0,
        transform: closed ? "translateY(100%)" : "translateY(0%)",
      }}
    >
      <div
        className={`absolute cursor-pointer`}
        style={{
          top: 5,
          right: 7,
          width: 13,
          height: 13,
        }}
        onClick={() => setClosed(true)}
      >
        <SmSolidPlus style={{ transform: "rotate(45deg)" }} />
      </div>
      <div className={`flex space-x-4 items-center`}>
        <p className={`bold`}>{message}</p>
        {button ? (
          <button
            className={`rounded-lg px-3 font-bold text-sm`}
            style={{
              backgroundColor: "#879EED",
              paddingTop: 3,
              paddingBottom: 3,
            }}
            {...props}
          >
            {button}
          </button>
        ) : null}
      </div>
    </div>
  );
};
