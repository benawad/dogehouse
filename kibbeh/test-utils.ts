import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import "mutationobserver-shim";

const customRender = (ui: any, options = {}) =>
  render(ui, {
    ...options,
  });

export * from "@testing-library/react";

export { customRender as render };
