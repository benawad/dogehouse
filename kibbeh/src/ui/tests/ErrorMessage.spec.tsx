import React from "react";

import { render, act } from "../../../test-utils";
import { ErrorMessage } from "../ErrorMessage";

jest.useFakeTimers();

describe("ErrorMessage", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const message = "some error message";
    const { getByTestId, getByText } = render(
      <ErrorMessage message={message} />
    );
    const component = getByTestId("error-message");
    const child = getByText(message);

    expect(component).toBeVisible();
    expect(child).toBeVisible();
  });

  it("should have translateY(0%) when rendered", () => {
    const message = "some error message";
    const { getByTestId } = render(
      <ErrorMessage message={message} autoClose={true} />
    );

    const component = getByTestId("error-message");

    expect(component.style.transform).toContain("translateY(0%)");
  });

  it("should have translateY(100%) after 7000ms if autoClose prop is true", () => {
    const message = "some error message";
    const { getByTestId } = render(
      <ErrorMessage message={message} autoClose={true} />
    );

    const component = getByTestId("error-message");

    act(() => {
      jest.runAllTimers();
    });
    expect(component.style.transform).toContain("translateY(100%)");
  });

  it("should have translateY(0%) after t > 7000ms if autoClose prop is false", () => {
    const message = "some error message";
    const { getByTestId } = render(
      <ErrorMessage message={message} autoClose={false} />
    );

    const component = getByTestId("error-message");

    act(() => {
      jest.advanceTimersByTime(8000);
    });

    expect(component.style.transform).toContain("translateY(0%)");
  });

  it("should have translateY(100%) after if the close button is clicked", () => {
    const message = "some error message";
    const { getByTestId } = render(
      <ErrorMessage message={message} autoClose={true} />
    );

    const component = getByTestId("error-message");
    const closeBtn = getByTestId("close-btn");
    closeBtn.click();

    expect(component.style.transform).toContain("translateY(100%)");
  });

  it("should match snapshot", () => {
    const { getByTestId } = render(<ErrorMessage />);
    const component = getByTestId("error-message");

    expect(component).toMatchSnapshot();
  });
});
