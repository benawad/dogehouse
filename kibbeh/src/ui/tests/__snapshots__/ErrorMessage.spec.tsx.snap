// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage should match snapshot 1`] = `
<div
  class="rounded-t-8 w-full fixed items-center justify-center text-button transition-transform duration-300"
  data-testid="error-message"
  style="height: 50px; background-color: rgb(85, 117, 231); bottom: 0px; left: 0px; transform: translateY(0%);"
>
  <div
    class="absolute cursor-pointer"
    data-testid="close-btn"
    style="top: 5px; right: 7px; width: 13px; height: 13px;"
  >
    <svg
      fill="currentColor"
      height="16"
      style="transform: rotate(45deg);"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clip-path="url(#clip12)"
      >
        <path
          clip-rule="evenodd"
          d="M8 0C8.55228 0 9 0.447715 9 1V15C9 15.5523 8.55228 16 8 16C7.44772 16 7 15.5523 7 15V1C7 0.447715 7.44772 0 8 0Z"
          fill="#DEE3EA"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M0.000976562 8C0.000976562 7.44772 0.448692 7 1.00098 7H15.001C15.5533 7 16.001 7.44772 16.001 8C16.001 8.55228 15.5533 9 15.001 9H1.00098C0.448692 9 0.000976562 8.55228 0.000976562 8Z"
          fill="#DEE3EA"
          fill-rule="evenodd"
        />
      </g>
      <defs>
        <clippath
          id="clip12"
        >
          <rect
            fill="white"
            height="16"
            width="16"
          />
        </clippath>
      </defs>
    </svg>
  </div>
  <div
    class="flex space-x-4 items-center"
  >
    <p
      class="bold"
    >
      An error has occurred
    </p>
    <button
      class="rounded-lg px-3 font-bold text-sm"
      style="background-color: rgb(135, 158, 237); padding-top: 3px; padding-bottom: 3px;"
    >
      Refresh
    </button>
  </div>
</div>
`;
